ksql-engine/src/main/java/io/confluent/ksql/function/UdfCompiler.java

```diff
@@ -148,28 +148,25 @@ public static UdfInvoker compile(final Method method, final ClassLoader loader)
               UDAF_PACKAGE + generatedClassName);
       final UdfArgSupplier evaluator = (UdfArgSupplier)
           scriptEvaluator.createFastEvaluator("return new " + generatedClassName
-                  + "(args, returnType, metrics);",
-              UdfArgSupplier.class, new String[]{"args", "returnType", "metrics"});
+                  + "(args, aggregateType, outputType, metrics);",
+              UdfArgSupplier.class, new String[]{"args", "aggregateType", "outputType", "metrics"});
 
-      final Schema argSchema = paramSchema.isEmpty()
-          ? UdfUtil.getSchemaFromType(valueAndAggregateTypes.left)
-          : SchemaConverters.sqlToConnectConverter()
-              .toConnectSchema(TypeContextUtil.getType(paramSchema).getSqlType());
+      final Schema argSchema = getSchemaFromType(inputOutputAndAggregateTypes[0], inputSchema);
       final List<Schema> args = Collections.singletonList(argSchema);
 
-      final Schema returnValue = returnSchema.isEmpty()
-          ? SchemaUtil.ensureOptional(UdfUtil.getSchemaFromType(valueAndAggregateTypes.right))
-          : SchemaConverters.sqlToConnectConverter()
-              .toConnectSchema(TypeContextUtil.getType(returnSchema).getSqlType());
+      final Schema aggregateValue = getSchemaFromType(inputOutputAndAggregateTypes[1],
+                                                      aggregateSchema);
 
-      return evaluator.apply(args, returnValue, metrics);
+      final Schema returnValue = getSchemaFromType(inputOutputAndAggregateTypes[2], outputSchema);
+
+      return evaluator.apply(args, aggregateValue, returnValue, metrics);
     } catch (final Exception e) {
       throw new KsqlException("Failed to compile KSqlAggregateFunction for method='"
           + method.getName() + "' in class='" + method.getDeclaringClass() + "'", e);
     }
   }
 
-  private static Pair<Type, Type> getValueAndAggregateTypes(
+  private static Type[] getInputOuputAndAggregateTypes(
```

Comment:
instead of an array, we should create a strongly typed inner class for this:
```java
class UdafTypes {
  Type inputType;
  Type aggregateType;
  Type outputType;
}
```